src/thunderbots/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,272 @@
+#include "ai/passing/pass_generator.h"
+
+#include <algorithm>
+#include <numeric>
+
+#include "ai/passing/evaluation.h"
+#include "pass_generator.h"
+
+
+using namespace AI::Passing;
+using namespace Util::DynamicParameters::AI::Passing;
+
+PassGenerator::PassGenerator(double min_reasonable_pass_quality)
+    : min_reasonable_pass_quality(min_reasonable_pass_quality),
+      optimizer(optimizer_param_weights),
+      best_known_pass(std::nullopt),
+      target_region(std::nullopt),
+      in_destructor(false)
+{
+    // Generate the initial set of passes
+    passes_to_optimize = generatePasses(num_passes_to_optimize.value());
+
+    // Start the thread to do the pass generation in the background
+    // The lambda expression here is needed so that we can call
+    // `continuouslyGeneratePasses()`, which is not a static function
+    pass_generation_thread =
+        std::thread([this]() { return continuouslyGeneratePasses(); });
+}
+
+void PassGenerator::setWorld(World world)
+{
+    // Take ownership of the world for the duration of this function
+    std::lock_guard<std::mutex> world_lock(world_mutex);
+
+    // Update the world
+    this->world = std::move(world);
```

Comment:
I'm not familiar with the rest of the structure for 'world', but I just want to double check that this won't cause threading issues. In the old AI, world was an object that contained pointers or references to the underlying data structures. So we chose to do a deep copy of the information we needed so we didn't need to worry about things being modified and read at the same time. Is this a non-issue with the new AI / ROS structure?

Reply:
@MathewMacDougall to confirm, but I don't believe we hold anything by pointer in the World object or it's members

Reply:
So will the world object used by the passing thread be completely unchanged except when setWorld() is called? And if that is the case now can it be guaranteed for future versions of the World class?

I don't see any immediate issues if it is changing in parallel, but this is the sort of thing that could lead to those super rare threading bugs that are impossible to debug

Reply:
Yeah, that's correct. Most of the system makes this assumption, it was a specific decision made early on to make things easier to reason about.

Reply:
I definitely agree with that design decision. Is there a way to guarantee that the copied world object has no links to the original? Either through class design, tests, or just a bunch of comments warning about how World and the objects it owns can be designed

The scenario I'm concerned about is if 2 years from now someone accidentally adds an array to one of the objects owned by world, and then each time world gets copied both objects actually point to the same memory

Reply:
Great point. Added the following comment to the javadoc comment for the `World` class:
```
 * WARNING: This class should _never_ hold any data that is pointed to anywhere else. This
 * means no raw pointers, no shared pointers, no shared references. This is because in
 * some cases we copy World's over multiple threads where having a shared member
 * between multiple instances of a World on multiple threads could result in
 * data corruption that would lead to absurdly hard to track bugs. YOU HAVE BEEN WARNED.
```

Reply:
I can confirm that the `World` currently doesn't hold references/pointers to any of its members, and this is not planned for the future. I agree with the comment that was added, and think we should be safe from threading issues.

Reply:
Looks good